src/awkward/contents/content.py

```diff
@@ -762,41 +761,23 @@ def _merging_strategy(
 
         head = [self]
         tail = []
-        i = 0
-
-        while i < len(others):
-            other = others[i]
-            if isinstance(
-                other,
-                (
-                    ak.contents.IndexedArray,
-                    ak.contents.IndexedOptionArray,
-                    ak.contents.ByteMaskedArray,
-                    ak.contents.BitMaskedArray,
-                    ak.contents.UnmaskedArray,
-                    ak.contents.UnionArray,
-                ),
-            ):
+
+        it_others = iter(others)
+
+        for other in it_others:
+            if other.is_indexed or other.is_option or other.is_union:
+                tail.append(other)
+                tail.extend(it_others)
                 break
             else:
                 head.append(other)
-            i = i + 1
-
-        while i < len(others):
-            tail.append(others[i])
-            i = i + 1
-
-        if any(isinstance(x.backend.nplike, TypeTracer) for x in head + tail):
-            head = [
-                x if isinstance(x.backend.nplike, TypeTracer) else x.to_typetracer()
-                for x in head
-            ]
-            tail = [
-                x if isinstance(x.backend.nplike, TypeTracer) else x.to_typetracer()
-                for x in tail
-            ]
-
-        return (head, tail)
+
+        if any(x.backend.nplike.known_data for x in head + tail) and not all(
+            x.backend.nplike.known_data for x in head + tail
+        ):
+            raise RuntimeError
```

Comment:
This could be left in forever as an assertion. (I hope I correctly inverted the logic: `not (any(...) and not all(...))` → `not any(...) or all(...)`.)

```suggestion
        assert not any(x.backend.nplike.known_data for x in head + tail) or all(
            x.backend.nplike.known_data for x in head + tail
        )
```

I can't believe I was paranoid enough to do this, but here's proof:

```python
>>> def f1(a, b):
...     return not(any(a) and not all(b))
... 
>>> def f2(a, b):
...     return not any(a) or all(b)
... 
>>> def make_a():
...     for i in range(2**4):
...         yield [x != "0" for x in format(i, "#06b")[-4:]]
... 
>>> from itertools import product
>>> for a, b in product(make_a(), make_a()):
...     assert f1(a, b) == f2(a, b)
... 
```

Reply:
Ha, I think binary ops are a perfect example of where one person's readable is not anothers! I'm happy to make the change, and it looks correct.